# Final Project Submission

Please fill out:
* Student name: Kristen Davis
* Student pace: Full Time
* Scheduled project review date/time: 11/16 @ 11:30
* Instructor name: Rafael Carrasco
* Blog post URL:


In [ ]:
https://rstudio-pubs-static.s3.amazonaws.com/339668_006f4906390e41cea23b3b786cc0230a.html

## Set Up & Data Initialization 

In [238]:
#Libraries 
import pandas as pd    

import matplotlib.pyplot as plt 
import seaborn as sns
import plotly.express as px 
import plotly.graph_objects as go 
from plotly.subplots import make_subplots  
%matplotlib inline


import plotly.io as pio
pio.renderers.default='notebook'

In [239]:
#Functions
def explore_df(df): 
    """Input: Takes in a DataFrame 
    Returns: shape, column names, non-null count, dtype, general statistical description, column values counts"""
    print(df.shape)
    print(df.isna().sum())
    print(list(df.columns)) 
    print(df.info()) 
    print(df.describe()) 
    return[df[col].value_counts() for col in df.columns]

### Load Data Sets

In [240]:
#Data for this contest downloaded from datadriven.org
test_set_values = pd.read_csv('/Users/kristen/Downloads/702ddfc5-68cd-4d1d-a0de-f5f566f76d91.csv') 
training_set_labels = pd.read_csv('/Users/kristen/Downloads/0bf8bc6e-30d0-4c50-956a-603fc693d966.csv')   
training_set_values = pd.read_csv('/Users/kristen/Downloads/4910797b-ee55-40a7-8668-10efd5c1b960.csv')

In [241]:
#Merge training set
df = pd.merge(training_set_labels, training_set_values, on="id") 

In [242]:
#Explore Data  
explore_df(df=df)

(59400, 41)
id                           0
status_group                 0
amount_tsh                   0
date_recorded                0
funder                    3635
gps_height                   0
installer                 3655
longitude                    0
latitude                     0
wpt_name                     0
num_private                  0
basin                        0
subvillage                 371
region                       0
region_code                  0
district_code                0
lga                          0
ward                         0
population                   0
public_meeting            3334
recorded_by                  0
scheme_management         3877
scheme_name              28166
permit                    3056
construction_year            0
extraction_type              0
extraction_type_group        0
extraction_type_class        0
management                   0
management_group             0
payment                      0
payment_type               

[2047     1
 72310    1
 49805    1
 51852    1
 62091    1
         ..
 46396    1
 36155    1
 34106    1
 38200    1
 0        1
 Name: id, Length: 59400, dtype: int64,
 functional                 32259
 non functional             22824
 functional needs repair     4317
 Name: status_group, dtype: int64,
 0.0         41639
 500.0        3102
 50.0         2472
 1000.0       1488
 20.0         1463
             ...  
 8500.0          1
 6300.0          1
 220.0           1
 138000.0        1
 12.0            1
 Name: amount_tsh, Length: 98, dtype: int64,
 2011-03-15    572
 2011-03-17    558
 2013-02-03    546
 2011-03-14    520
 2011-03-16    513
              ... 
 2011-09-17      1
 2011-09-15      1
 2011-09-25      1
 2002-10-14      1
 2011-09-13      1
 Name: date_recorded, Length: 356, dtype: int64,
 Government Of Tanzania    9084
 Danida                    3114
 Hesawa                    2202
 Rwssp                     1374
 World Bank                1349
                   

Structure:
* The data has 59,400 unique enteries with 41 features(columns)  

Nan Values: 
* 3,635 values are missing in the funder & installer category (6% of data missing) 
* 371 values are missing from the subvillage category (>1% of data missing) 
* 3,334 values are missing from the public meeting category (5% of data missing)  
* 3,877 values are missing from the scheme management category (6.5% of data missing)
* 28,166 values are missing from the scheme name category (47% of data missing) 
* 3,056 values are missing from the permit category (5% of data missing) 

### Column Types

In [243]:
df.columns

Index(['id', 'status_group', 'amount_tsh', 'date_recorded', 'funder',
       'gps_height', 'installer', 'longitude', 'latitude', 'wpt_name',
       'num_private', 'basin', 'subvillage', 'region', 'region_code',
       'district_code', 'lga', 'ward', 'population', 'public_meeting',
       'recorded_by', 'scheme_management', 'scheme_name', 'permit',
       'construction_year', 'extraction_type', 'extraction_type_group',
       'extraction_type_class', 'management', 'management_group', 'payment',
       'payment_type', 'water_quality', 'quality_group', 'quantity',
       'quantity_group', 'source', 'source_type', 'source_class',
       'waterpoint_type', 'waterpoint_type_group'],
      dtype='object')

In [244]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 59400 entries, 0 to 59399
Data columns (total 41 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   id                     59400 non-null  int64  
 1   status_group           59400 non-null  object 
 2   amount_tsh             59400 non-null  float64
 3   date_recorded          59400 non-null  object 
 4   funder                 55765 non-null  object 
 5   gps_height             59400 non-null  int64  
 6   installer              55745 non-null  object 
 7   longitude              59400 non-null  float64
 8   latitude               59400 non-null  float64
 9   wpt_name               59400 non-null  object 
 10  num_private            59400 non-null  int64  
 11  basin                  59400 non-null  object 
 12  subvillage             59029 non-null  object 
 13  region                 59400 non-null  object 
 14  region_code            59400 non-null  int64  
 15  di

In [245]:
#itrative cell to look a unique values to identify type 
df['permit'].unique()

array([False, True, nan], dtype=object)

* <b>Reference</b> - 'id', 'date_recorded', 'recorded_by'


* <b>Categorical</b> - 'status_group', 'funder', 'installer', 'wpt_name', 'basin', 'subvillage', 'region', 'region_code', 'district_code', 'lga', 'ward', 'public_meeting', 'scheme_management', 'scheme_name', 'permit', 'construction_year', 'extraction_type', 'extraction_type_group', 'extraction_type_class', 'management', 'management_group', 'payment', 'payment_type', 'water_quality', 'quality_group', 'quantity', 'quantity_group', 'source', 'source_type', 'source_class', 'waterpoint_type', 'waterpoint_type_group'


* <b>Continous</b> - 'amount_tsh', 'gps_height', 'longitude', 'latitude', 'num_private', 'population'

## Investigate Each Reference Column 

### id 

In [246]:
df['id']

0        69572
1         8776
2        34310
3        67743
4        19728
         ...  
59395    60739
59396    27263
59397    37057
59398    31282
59399    26348
Name: id, Length: 59400, dtype: int64

59400 unique identifing ids correlated to a unique water pump. 

### date_recorded

In [247]:
df[['date_recorded']] 
print(df[['date_recorded']].isna().sum()) 
df[['date_recorded']].info

date_recorded    0
dtype: int64


<bound method DataFrame.info of       date_recorded
0        2011-03-14
1        2013-03-06
2        2013-02-25
3        2013-01-28
4        2011-07-13
...             ...
59395    2013-05-03
59396    2011-05-07
59397    2011-04-11
59398    2011-03-08
59399    2011-03-23

[59400 rows x 1 columns]>

In [251]:
df['date_recorded'] = pd.to_datetime(df['date_recorded'])  

In [252]:
dr_df = df.groupby('date_recorded').count().reset_index()  
dr_df

date_recorded   id  status_group  amount_tsh  funder  gps_height  \
0      2002-10-14    1             1           1       1           1   
1      2004-01-07    1             1           1       1           1   
2      2004-03-01    4             4           4       4           4   
3      2004-03-06    1             1           1       1           1   
4      2004-04-01    1             1           1       1           1   
..            ...  ...           ...         ...     ...         ...   
351    2013-11-02   24            24          24      24          24   
352    2013-11-03  194           194         194     194         194   
353    2013-12-01    1             1           1       1           1   
354    2013-12-02   33            33          33      33          33   
355    2013-12-03  240           240         240     239         240   

     installer  longitude  latitude  wpt_name  ...  payment_type  \
0            1          1         1         1  ...             1   
1            1          1         1         1  ...             1   
2            4          4         4         4  ...             4   
3            1          1         1         1  ...             1   
4            1          1         1         1  ...             1   
..         ...        ...       ...       ...  ...           ...   
351         24         24        24        24  ...            24   
352        194        194       194       194  ...           194   
353          1          1         1         1  ...             1   
354         33         33        33        33  ...            33   
355        240        240       240       240  ...           240   

     water_quality  quality_group  quantity  quantity_group  source  \
0                1              1         1               1       1   
1                1              1         1               1       1   
2                4              4         4               4       4   
3                1              1         1               1       1   
4                1              1         1               1       1   
..             ...            ...       ...             ...     ...   
351             24             24        24              24      24   
352            194            194       194             194     194   
353              1              1         1               1       1   
354             33             33        33              33      33   
355            240            240       240             240     240   

     source_type  source_class  waterpoint_type  waterpoint_type_group  
0              1             1                1                      1  
1              1             1                1                      1  
2              4             4                4                      4  
3              1             1                1                      1  
4              1             1                1                      1  
..           ...           ...              ...                    ...  
351           24            24               24                     24  
352          194           194              194                    194  
353            1             1                1                      1  
354           33            33               33                     33  
355          240           240              240                    240  

[356 rows x 41 columns]

In [253]:
fig = px.line(dr_df, x='date_recorded', y='id')
fig.show()

* Date data was recorded for each pump.  
* Non NaN values  
* The majority of the obeservations after 2010 with the largest number of data (572) recorded in March of 2011

### recorded_by

In [263]:
df[['recorded_by']] 
print(df[['recorded_by']].isna().sum()) 
df[['recorded_by']].info() 
print(df['recorded_by'].unique())

recorded_by    0
dtype: int64
<class 'pandas.core.frame.DataFrame'>
Int64Index: 59400 entries, 0 to 59399
Data columns (total 1 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   recorded_by  59400 non-null  object
dtypes: object(1)
memory usage: 928.1+ KB
['GeoData Consultants Ltd']


* All values/ observations recorded by GeoData Consultants Ltd
* No NaN values

## Investigate Each Categorical Column

### status_group

In [269]:
df[['status_group']] 
print(df[['status_group']].isna().sum())
df[['status_group']].info() 
print(df['status_group'].unique())

status_group    0
dtype: int64
<class 'pandas.core.frame.DataFrame'>
Int64Index: 59400 entries, 0 to 59399
Data columns (total 1 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   status_group  59400 non-null  object
dtypes: object(1)
memory usage: 928.1+ KB
['functional' 'non functional' 'functional needs repair']


In [270]:
#Length of each new dataset
print(len(functional))
print(len(non_functional))
print(len(functional_needs_repair))

32259
22824
4317


In [271]:
labels = ['Functional', 'Non Functional', 'Functional Needs Repair']
values = [32259, 22824, 4317]

fig = go.Figure(data=[go.Pie(labels=labels, values=values)]) 
fig.update_traces(textposition='inside', textinfo='percent+label', title='Water Pump Status')
fig.show()

* These are the categories that ultimately the model is attempting to sort into. Functional, non functional, and funtional needs repair 
* 54.3% Functional Pumps/ 38.4% Non Functional/ 7.27% Functional Needs Repair 

61.57% of the pumps in the data set are functional

### funder

In [273]:
df[['funder']] 
print(df[['funder']].isna().sum())
df[['funder']].info() 

funder    3635
dtype: int64
<class 'pandas.core.frame.DataFrame'>
Int64Index: 59400 entries, 0 to 59399
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   funder  55765 non-null  object
dtypes: object(1)
memory usage: 928.1+ KB
['Roman' 'Grumeti' 'Lottery Club' ... 'Dina' 'Brown' 'Samlo']


In [276]:
print(df[['funder']].isna().sum())

funder    3635
dtype: int64


In [275]:
df.groupby('funder').count()

id  status_group  amount_tsh  date_recorded  gps_height  \
funder                                                                          
0                    777           777         777            777         777   
A/co Germany          15            15          15             15          15   
Aar                   33            33          33             33          33   
Abas Ka                1             1           1              1           1   
Abasia                43            43          43             43          43   
...                  ...           ...         ...            ...         ...   
Zao                    1             1           1              1           1   
Zao Water Spring       3             3           3              3           3   
Zao Water Spring X     1             1           1              1           1   
Zinduka                1             1           1              1           1   
Zingibali Secondary    1             1           1              1           1   

                     installer  longitude  latitude  wpt_name  num_private  \
funder                                                                       
0                          777        777       777       777          777   
A/co Germany                15         15        15        15           15   
Aar                         32         33        33        33           33   
Abas Ka                      1          1         1         1            1   
Abasia                      43         43        43        43           43   
...                        ...        ...       ...       ...          ...   
Zao                          1          1         1         1            1   
Zao Water Spring             3          3         3         3            3   
Zao Water Spring X           1          1         1         1            1   
Zinduka                      1          1         1         1            1   
Zingibali Secondary          1          1         1         1            1   

                     ...  payment_type  water_quality  quality_group  \
funder               ...                                               
0                    ...           777            777            777   
A/co Germany         ...            15             15             15   
Aar                  ...            33             33             33   
Abas Ka              ...             1              1              1   
Abasia               ...            43             43             43   
...                  ...           ...            ...            ...   
Zao                  ...             1              1              1   
Zao Water Spring     ...             3              3              3   
Zao Water Spring X   ...             1              1              1   
Zinduka              ...             1              1              1   
Zingibali Secondary  ...             1              1              1   

                     quantity  quantity_group  source  source_type  \
funder                                                               
0                         777             777     777          777   
A/co Germany               15              15      15           15   
Aar                        33              33      33           33   
Abas Ka                     1               1       1            1   
Abasia                     43              43      43           43   
...                       ...             ...     ...          ...   
Zao                         1               1       1            1   
Zao Water Spring            3               3       3            3   
Zao Water Spring X          1               1       1            1   
Zinduka                     1               1       1            1   
Zingibali Secondary         1               1       1            1   

                     source_class  waterpoint_type  waterpoint_type_group  
funder                       

In [284]:
df.loc[df['funder'] == '0']

id    status_group  amount_tsh date_recorded funder  gps_height  \
145    28669      functional        50.0    2013-03-21      0          12   
152    60983  non functional         0.0    2013-03-16      0         -15   
393    39749  non functional         0.0    2013-03-18      0          28   
417    15832  non functional        50.0    2013-03-22      0          30   
428    50233  non functional         0.0    2013-03-12      0          30   
...      ...             ...         ...           ...    ...         ...   
59237   2138  non functional        59.0    2013-03-19      0          81   
59243   3396      functional        50.0    2013-03-16      0         -20   
59276  62818      functional        50.0    2013-03-21      0          18   
59351  55322      functional        50.0    2013-03-18      0         -19   
59387  26640      functional       100.0    2013-03-12      0          25   

      installer  longitude  latitude       wpt_name  ...  payment_type  \
145           0  39.185710 -6.892593       Chekanao  ...    per bucket   
152           0  39.527114 -6.988748      Msikitini  ...     never pay   
393           0  39.159887 -6.902548   Kwa Chambuso  ...       unknown   
417           0  39.178404 -6.938013     Ccm Kivule  ...    per bucket   
428           0  39.178849 -6.973206  Ofisi Ya Kata  ...     never pay   
...         ...        ...       ...            ...  ...           ...   
59237         0  39.119109 -6.898919           Kata  ...    per bucket   
59243         0  39.524021 -6.984802   Kwa Mariwala  ...    per bucket   
59276         0  39.183790 -6.897566   Kwa Mkunduge  ...    per bucket   
59351         0  39.534599 -7.088183      Kwa China  ...    per bucket   
59387         0  39.176480 -6.957098     Kwa Maliba  ...    per bucket   

         water_quality quality_group      quantity  quantity_group  \
145               soft          good        enough          enough   
152              salty         salty           dry             dry   
393            unknown       unknown       unknown         unknown   
417               soft          good        enough          enough   
428            unknown       unknown           dry             dry   
...                ...           ...           ...             ...   
59237  salty abandoned         salty  insufficient    insufficient   
59243             soft          good        enough          enough   
59276             soft          good        enough          enough   
59351             soft          good        enough          enough   
59387             soft          good        enough          enough   

             source   source_type source_class     waterpoint_type  \
145     machine dbh      borehole  groundwater  communal standpipe   
152     machine dbh      borehole  groundwater               other   
393     machine dbh      borehole  groundwater               other   
417     machine dbh      borehole  groundwater  communal standpipe   
428    shallow well  shallow well  groundwater               other   
...             ...           ...          ...                 ...   
59237   machine dbh      borehole  groundwater  communal standpipe   
59243  shallow well  shallow well  groundwater           hand pump   
59276   machine dbh      borehole  groundwater  communal standpipe   
59351   machine dbh      borehole  groundwater  communal standpipe   
59387   machine dbh      borehole  groundwater  communal standpipe   

      waterpoint_type_group  
145      communal standpipe  
152                   other  
393                   other  
417      communal standpipe  
428                   other  
...                     ...  
59237    communal standpipe  
59243             hand pump  
59276    communal standpipe  
59351    communal standpipe  
59387    communal standpipe  

[777 rows x 41 columns]

* 3635 values are missing funder 
* Many of these enteries appear to be duplicates ie "Zao Water Spring/ Zao Water Spring X/ Zao" for example 
* Rows that have a 0 listed as a funder also have a 0 listed as an installer

### installer

### wpt_name

### basin

### subvillage

### region

### region_code

### district_code

### lga

In [ ]:
'ward', 'public_meeting', 'scheme_management', 'scheme_name', 'permit', 'construction_year', 'extraction_type', 'extraction_type_group', 'extraction_type_class', 'management', 'management_group', 'payment', 'payment_type', 'water_quality', 'quality_group', 'quantity', 'quantity_group', 'source', 'source_type', 'source_class', 'waterpoint_type', 'waterpoint_type_group'

## Investigate Each Continous Column 

### amount_tsh

### gps_height

### longitude

### latitude

### num_private

### population

### Column Values & Duplicates

In [175]:
#column names 
df.columns

Index(['id', 'status_group', 'amount_tsh', 'date_recorded', 'funder',
       'gps_height', 'installer', 'longitude', 'latitude', 'wpt_name',
       'num_private', 'basin', 'subvillage', 'region', 'region_code',
       'district_code', 'lga', 'ward', 'population', 'public_meeting',
       'recorded_by', 'scheme_management', 'scheme_name', 'permit',
       'construction_year', 'extraction_type', 'extraction_type_group',
       'extraction_type_class', 'management', 'management_group', 'payment',
       'payment_type', 'water_quality', 'quality_group', 'quantity',
       'quantity_group', 'source', 'source_type', 'source_class',
       'waterpoint_type', 'waterpoint_type_group'],
      dtype='object')

In [176]:
#Visually inspect extraction columns
df[['extraction_type_class', 'extraction_type', 'extraction_type_group']]

extraction_type_class extraction_type extraction_type_group
0                   gravity         gravity               gravity
1                   gravity         gravity               gravity
2                   gravity         gravity               gravity
3               submersible     submersible           submersible
4                   gravity         gravity               gravity
...                     ...             ...                   ...
59395               gravity         gravity               gravity
59396               gravity         gravity               gravity
59397              handpump          swn 80                swn 80
59398              handpump     nira/tanira           nira/tanira
59399              handpump     nira/tanira           nira/tanira

[59400 rows x 3 columns]

In [177]:
#Visually inpsect payment columns
df[['payment_type', 'payment']]

payment_type                payment
0         annually           pay annually
1        never pay              never pay
2       per bucket         pay per bucket
3        never pay              never pay
4        never pay              never pay
...            ...                    ...
59395   per bucket         pay per bucket
59396     annually           pay annually
59397      monthly            pay monthly
59398    never pay              never pay
59399   on failure  pay when scheme fails

[59400 rows x 2 columns]

In [169]:
raw_df.loc[raw_df['water_quality'] == 'fluoride abandoned'] 
#raw_df['quality_group'].value_counts()

id    status_group  amount_tsh date_recorded  \
6793   10874  non functional         0.0    2013-03-19   
7372   46756  non functional         0.0    2012-11-10   
10759  33573  non functional         0.0    2011-03-16   
14432  33126      functional         0.0    2012-11-12   
14595  31756  non functional         0.0    2012-10-22   
15872  20993  non functional         0.0    2011-03-05   
18015  60886      functional         0.0    2013-03-07   
20732  58888  non functional         0.0    2013-02-17   
22165  70990      functional         0.0    2013-02-10   
29148  46721      functional         0.0    2011-03-17   
31730  32143  non functional         0.0    2012-10-19   
32950  40880  non functional         0.0    2011-03-05   
33201  38144  non functional         0.0    2013-01-29   
33395  54363      functional         0.0    2013-02-05   
34135  22207      functional         0.0    2012-11-12   
38992  60616  non functional         0.0    2013-03-19   
50513  14139  non functional         0.0    2012-10-21   

                       funder  gps_height          installer  longitude  \
6793                        0         -30                  0  39.363450   
7372             World Vision           0       World vision  33.931583   
10759  Government Of Tanzania         299                DWE  36.137635   
14432             Netherlands           0                DWE  34.020473   
14595                    Wsdp           0  AQUARMAN DRILLERS  34.107893   
15872                   Amref          18              AMREF  39.057837   
18015            World Vision         952       World Vision  35.735400   
20732                    Lamp        1524                DWE  35.594111   
22165                    Dwsp           0                DWE   0.000000   
29148  Government Of Tanzania         293                DWE  36.135270   
31730  Government Of Tanzania           0         Government  33.809141   
32950                  Rished         -17                DWE  38.946685   
33201  Government Of Tanzania        1437                DWE  30.713918   
33395                     Wvt           0                WVT   0.000000   
34135            World Vision           0       World vision  34.021197   
38992                       0         -19                  0  39.430507   
50513                   Rwssp           0                DWE  32.390626   

           latitude               wpt_name  ...  payment_type  \
6793  -6.894742e+00             Kwa Mbonde  ...     never pay   
7372  -3.486847e+00          Kwa Kazengele  ...         other   
10759 -8.942505e+00                 Sokoni  ...       monthly   
14432 -3.582536e+00           Busulwandulu  ...     never pay   
14595 -3.417657e+00                   None  ...     never pay   
15872 -7.396576e+00          Kwa Mangwenya  ...     never pay   
18015 -3.877415e+00               Kazaroho  ...     never pay   
20732 -4.340430e+00            Kwa Mathias  ...     never pay   
22165 -2.000000e-08                Kwamagu  ...       unknown   
29148 -8.942596e+00              Msikitini  ...     never pay   
31730 -3.441827e+00            Pius Mwandu  ...         other   
32950 -5.379934e+00          Kwa Mzee Juma  ...     never pay   
33201 -3.603296e+00  Kuzera Primary School  ...       unknown   
33395 -2.000000e-08             Njia Panda  ...     never pay   
34135 -3.583229e+00                Msituni  ...         other   
38992 -6.940968e+00                Shuleni  ...     never pay   
50513 -4.031776e+00                Kangeme  ...       unknown   

            water_quality quality_group      quantity  quantity_group  \
6793   fluoride abandoned      fluoride  insufficient    insufficient   
7372   fluoride abandoned      fluoride        enough          enough   
10759  fluoride abandoned      fluoride  insufficient    insufficient   
14432  fluoride abandoned      fluoride        enough          enough   
14595  fluoride abandoned      fluoride        enough          enough   
15872

Duplicate/ Similar Columns: 
* The ('extraction_type', 'extraction_type_group','extraction_type_class') are larger subgroupings/ generalizations of the data  
* The ('payment_type & 'payment') columns are completely duplicated I am going to drop payment_type 
* The ('water_quality' & 'quality_group') are very similar:  good correlates to "soft" / salty

# Data Cleaning 

## Dropping Columns

In [153]:
df1 = df.drop(['water_quality', 'payment_type'], axis=1)

replace 0 in funder and installer to be unknown

In [141]:
#df.loc[(df['funder'] == '0') & (df['installer'] == '0')]  

#777 of the rows with no funder also have no installer  listed as well

Index(['id', 'status_group', 'amount_tsh', 'date_recorded', 'funder',
       'gps_height', 'installer', 'longitude', 'latitude', 'wpt_name',
       'num_private', 'basin', 'subvillage', 'region', 'region_code',
       'district_code', 'lga', 'ward', 'population', 'public_meeting',
       'recorded_by', 'scheme_management', 'scheme_name', 'permit',
       'construction_year', 'extraction_type', 'extraction_type_group',
       'extraction_type_class', 'management', 'management_group', 'payment',
       'payment_type', 'water_quality', 'quality_group', 'quantity',
       'quantity_group', 'source', 'source_type', 'source_class',
       'waterpoint_type', 'waterpoint_type_group'],
      dtype='object')

## EDA 

In [6]:
#divide data by status_group
functional = df.loc[df['status_group'] == 'functional'] 
non_functional = df.loc[df['status_group'] == 'non functional'] 
functional_needs_repair = df.loc[df['status_group'] == 'functional needs repair']  

In [52]:
m = folium.Map(location=[6.3690, 34.8888]) 
m

### Question: What is similar about functioning pumps which are producing enough good water? 

In [39]:
good_enough_functional = functional.loc[(functional['quantity'] == 'enough') & (functional['quality_group'] == 'good')] 
good_enough_functional.head()

id status_group  amount_tsh date_recorded                   funder  \
0   69572   functional      6000.0    2011-03-14                    Roman   
2   34310   functional        25.0    2013-02-25             Lottery Club   
9   46144   functional         0.0    2011-08-03              Isingiro Ho   
12  36957   functional         0.0    2012-10-14             World Vision   
13  50495   functional         0.0    2013-03-15  Lawatefuka Water Supply   

    gps_height             installer  longitude  latitude  \
0         1390                 Roman  34.938093 -9.856322   
2          686          World vision  37.460664 -3.821329   
9            0               Artisan  30.626991 -1.257051   
12           0          World vision  33.798106 -3.290194   
13        1368  Lawatefuka water sup  37.092574 -3.181783   

                wpt_name  ...  payment_type water_quality quality_group  \
0                   none  ...      annually          soft          good   
2            Kwa Mahundi  ...    per bucket          soft          good   
9                Kwapeto  ...     never pay          soft          good   
12                 Pamba  ...         other          soft          good   
13  Kwa John Izack Mmari  ...       monthly          soft          good   

   quantity  quantity_group        source   source_type source_class  \
0    enough          enough        spring        spring  groundwater   
2    enough          enough           dam           dam      surface   
9    enough          enough  shallow well  shallow well  groundwater   
12   enough          enough  shallow well  shallow well  groundwater   
13   enough          enough        spring        spring  groundwater   

                waterpoint_type waterpoint_type_group  
0            communal standpipe    communal standpipe  
2   communal standpipe multiple    communal standpipe  
9                     hand pump             hand pump  
12                    hand pump             hand pump  
13           communal standpipe    communal standpipe  

[5 rows x 41 columns]

In [65]:
explore_df(df=good_enough_functional)

(19640, 41)
id                          0
status_group                0
amount_tsh                  0
date_recorded               0
funder                    739
gps_height                  0
installer                 743
longitude                   0
latitude                    0
wpt_name                    0
num_private                 0
basin                       0
subvillage                 42
region                      0
region_code                 0
district_code               0
lga                         0
ward                        0
population                  0
public_meeting            923
recorded_by                 0
scheme_management         678
scheme_name              7403
permit                    699
construction_year           0
extraction_type             0
extraction_type_group       0
extraction_type_class       0
management                  0
management_group            0
payment                     0
payment_type                0
water_quality               

[36862    1
 51948    1
 8929     1
 45795    1
 6886     1
         ..
 54656    1
 60807    1
 5512     1
 7561     1
 34817    1
 Name: id, Length: 19640, dtype: int64,
 functional    19640
 Name: status_group, dtype: int64,
 0.0         10872
 500.0        1710
 50.0         1122
 1000.0        751
 20.0          526
             ...  
 13000.0         1
 53.0            1
 590.0           1
 138000.0        1
 1300.0          1
 Name: amount_tsh, Length: 81, dtype: int64,
 2011-03-17    286
 2011-03-15    280
 2011-02-28    228
 2011-03-18    221
 2011-03-14    207
              ... 
 2004-06-01      1
 2011-09-18      1
 2011-08-20      1
 2013-12-01      1
 2012-12-11      1
 Name: date_recorded, Length: 313, dtype: int64,
 Government Of Tanzania           2131
 Danida                           1277
 Private Individual                542
 Germany Republi                   538
 Unicef                            507
                                  ... 
 Healt                    

#### Are there any trends in the funder in this category? 

In [67]:
funder_good_enough_functional = good_enough_functional.groupby('funder').count().reset_index()

In [70]:
funder_good_enough_functional_over_100 = funder_good_enough_functional.loc[funder_good_enough_functional['id'] > 100].sort_values('id', ascending=False)

In [73]:
fig = go.Figure(data=[go.Bar(x=funder_good_enough_functional_over_100['funder'], y=funder_good_enough_functional_over_100['id'])])
# Customize aspect
fig.update_traces(marker_color='rgb(158,202,225)', marker_line_color='rgb(8,48,107)',
                  marker_line_width=1.5, opacity=0.6)
fig.update_layout(title_text='Funders on Sufficently Functioning Water Pumps Producing Good Water')
fig.show()

#### What years were the pumps in this category made?

In [81]:
construction_year_good_enough_functional = good_enough_functional.groupby('construction_year').count().reset_index()

In [90]:
construction_year_good_enough_functional.sum()

construction_year        105310
id                        19640
status_group              19640
amount_tsh                19640
date_recorded             19640
funder                    18901
gps_height                19640
installer                 18897
longitude                 19640
latitude                  19640
wpt_name                  19640
num_private               19640
basin                     19640
subvillage                19598
region                    19640
region_code               19640
district_code             19640
lga                       19640
ward                      19640
population                19640
public_meeting            18717
recorded_by               19640
scheme_management         18962
scheme_name               12237
permit                    18941
extraction_type           19640
extraction_type_group     19640
extraction_type_class     19640
management                19640
management_group          19640
payment                   19640
payment_

In [93]:
x= [1960, 1970, 1980, 1990, 2000, 2010] 
y = [79, 804, 1397, 2617, 6604, 2276]
fig = go.Figure(data=[go.Bar(x=x, y=y, hovertext=['~0.1%', '~4%', '~7%', '~13%', '~34', '~12%'])])
# Customize aspect
fig.update_traces(marker_color='rgb(158,202,225)', marker_line_color='rgb(8,48,107)',
                  marker_line_width=1.5, opacity=0.6)
fig.update_layout(title_text='Decade of Construction on Sufficently Functioning Water Pumps Producing Good Water')
fig.show()

#### What is the breakdown of payment type in this category?

In [60]:
#group by payment method
payment_good_enough_functional = good_enough_functional.groupby('payment_type').count() 
#payment_good_enough_functional.reset_index(inplace=True) 

In [63]:
payment_good_enough_functional

payment_type    id  status_group  amount_tsh  date_recorded  funder  \
0     annually  2117          2117        2117           2117    2095   
1      monthly  3665          3665        3665           3665    3652   
2    never pay  6432          6432        6432           6432    5946   
3   on failure  1506          1506        1506           1506    1503   
4        other   389           389         389            389     387   
5   per bucket  3894          3894        3894           3894    3858   
6      unknown  1637          1637        1637           1637    1460   

   gps_height  installer  longitude  latitude  ...  payment  water_quality  \
0        2117       2091       2117      2117  ...     2117           2117   
1        3665       3651       3665      3665  ...     3665           3665   
2        6432       5953       6432      6432  ...     6432           6432   
3        1506       1503       1506      1506  ...     1506           1506   
4         389        386        389       389  ...      389            389   
5        3894       3856       3894      3894  ...     3894           3894   
6        1637       1457       1637      1637  ...     1637           1637   

   quality_group  quantity  quantity_group  source  source_type  source_class  \
0           2117      2117            2117    2117         2117          2117   
1           3665      3665            3665    3665         3665          3665   
2           6432      6432            6432    6432         6432          6432   
3           1506      1506            1506    1506         1506          1506   
4            389       389             389     389          389           389   
5           3894      3894            3894    3894         3894          3894   
6           1637      1637            1637    1637         1637          1637   

   waterpoint_type  waterpoint_type_group  
0             2117                   2117  
1             3665                   3665  
2             6432                   6432  
3             1506                   1506  
4              389                    389  
5             3894                   3894  
6             1637                   1637  

[7 rows x 41 columns]

In [64]:
fig = go.Figure(data=[go.Bar(x=payment_good_enough_functional['payment_type'], y=payment_good_enough_functional['id'],
            hovertext=['~11%', '~19%', '~33%', '~8%', '~2%', '~20%', '~8%'])])
# Customize aspect
fig.update_traces(marker_color='rgb(158,202,225)', marker_line_color='rgb(8,48,107)',
                  marker_line_width=1.5, opacity=0.6)
fig.update_layout(title_text='Payment Methods on Sufficently Functioning Water Pumps Producing Good Water')
fig.show()

### Question: What trends can be identified in pumps built by the Tanzanian Government?  
15% of all water pumps in this data set were funded for by the Tanzanian Governement

In [102]:
tg_pumps = df.loc[df['funder'] == 'Government Of Tanzania'] 
tg_pumps

id             status_group  amount_tsh date_recorded  \
25     22308  functional needs repair         0.0    2013-02-25   
28     19685           non functional         0.0    2013-03-05   
32     57238  functional needs repair         0.0    2013-03-19   
39        25           non functional         0.0    2013-02-11   
41     12796               functional       500.0    2011-03-12   
...      ...                      ...         ...           ...   
59367  73019               functional      2000.0    2011-03-09   
59375  50998               functional     40000.0    2011-08-03   
59377  43986           non functional         0.0    2012-10-23   
59391  44885           non functional         0.0    2013-08-03   
59392  40607           non functional         0.0    2011-04-15   

                       funder  gps_height         installer  longitude  \
25     Government Of Tanzania        1273               DWE  37.422751   
28     Government Of Tanzania        1443  District council  37.611126   
32     Government Of Tanzania        1481               DWE  37.492334   
39     Government Of Tanzania        1348               DWE  30.613305   
41     Government Of Tanzania        2469             Commu  33.927902   
...                       ...         ...               ...        ...   
59367  Government Of Tanzania        1977             Commu  34.338899   
59375  Government Of Tanzania          54               DWE  39.046942   
59377  Government Of Tanzania           0        Government  33.437012   
59391  Government Of Tanzania         540        Government  38.044070   
59392  Government Of Tanzania           0        Government  33.009440   

       latitude          wpt_name  ...  payment_type water_quality  \
25    -3.317536         Kwa Tukai  ...       unknown          soft   
28    -3.263526      Kwa Kibakaya  ...       monthly          soft   
32    -3.290627  Kwa Philip Lyimo  ...     never pay          soft   
39    -3.894561           Mkuyuni  ...       unknown          soft   
41    -9.204649              none  ...    on failure          soft   
...         ...               ...  ...           ...           ...   
59367 -9.565553              none  ...      annually          soft   
59375 -5.061494         Sekondari  ...       monthly          soft   
59377 -4.423594              Fpct  ...     never pay          soft   
59391 -4.272218               Kwa  ...     never pay          soft   
59392 -8.520888    Benard Charles  ...     never pay          soft   

      quality_group      quantity  quantity_group        source   source_type  \
25             good        enough          enough        spring        spring   
28             good  insufficient    insufficient        spring        spring   
32             good  insufficient    insufficient         river    river/lake   
39             good      seasonal        seasonal  shallow well  shallow well   
41             good        enough          enough        spring        spring   
...             ...           ...             ...           ...           ...   
59367          good        enough          enough        spring        spring   
59375          good        enough          enough         river    river/lake   
59377          good  insufficient    insufficient           dam           dam   
59391          good        enough          enough         river    river/lake   
59392          good        enough          enough        spring        spring   

      source_class              waterpoint_type waterpoint_type_group  
25     groundwater           communal standpipe    communal standpipe  
28     groundwater           communal standpipe    communal standpipe  
32         surface                        other                 other  
39     groundwater                    hand pump             hand pump  
41     groundwater           communal standpipe    communal standpipe  
...            ...                          ...                   ... 

#### What percentage of government pumps are functioning? 

In [118]:
tg_pumps.groupby('status_group').count()

id  amount_tsh  date_recorded  funder  gps_height  \
status_group                                                                   
functional               3720        3720           3720    3720        3720   
functional needs repair   701         701            701     701         701   
non functional           4663        4663           4663    4663        4663   

                         installer  longitude  latitude  wpt_name  \
status_group                                                        
functional                    3717       3720      3720      3720   
functional needs repair        701        701       701       701   
non functional                4662       4663      4663      4663   

                         num_private  ...  payment_type  water_quality  \
status_group                          ...                                
functional                      3720  ...          3720           3720   
functional needs repair          701  ...           701            701   
non functional                  4663  ...          4663           4663   

                         quality_group  quantity  quantity_group  source  \
status_group                                                               
functional                        3720      3720            3720    3720   
functional needs repair            701       701             701     701   
non functional                    4663      4663            4663    4663   

                         source_type  source_class  waterpoint_type  \
status_group                                                          
functional                      3720          3720             3720   
functional needs repair          701           701              701   
non functional                  4663          4663             4663   

                         waterpoint_type_group  
status_group                                    
functional                                3720  
functional needs repair                    701  
non functional                            4663  

[3 rows x 40 columns]

In [116]:
x = ['Functional', 'Functional Needs Repair', 'Non Functional']
y = [3720, 701, 4663]

# Use the hovertext kw argument for hover text
fig = go.Figure(data=[go.Bar(x=x, y=y,
            hovertext=['~41%', '~8%', '~51%'])])
# Customize aspect
fig.update_traces(marker_color='rgb(158,202,225)', marker_line_color='rgb(8,48,107)',
                  marker_line_width=1.5, opacity=0.6)
fig.update_layout(title_text='Government Build Water Pumps by Status')
fig.show()

#### Who installs the pumps the govement builds? 

In [121]:
installer_tg_pumps = tg_pumps.groupby('installer').count().reset_index()

In [122]:
fig = go.Figure(data=[go.Bar(x=installer_tg_pumps['installer'], y=installer_tg_pumps['id'])])
# Customize aspect
fig.update_traces(marker_color='rgb(158,202,225)', marker_line_color='rgb(8,48,107)',
                  marker_line_width=1.5, opacity=0.6)
fig.update_layout(title_text='Installers of Government Water Pumps')
fig.show()

# Working Line

In [8]:
df['water_quality'].value_counts()

soft                  50818
salty                  4856
unknown                1876
milky                   804
coloured                490
salty abandoned         339
fluoride                200
fluoride abandoned       17
Name: water_quality, dtype: int64

In [20]:
soft = df.loc[df['water_quality'] == 'soft']